espnet2/enh/layers/tcndenseunet.py

```diff
@@ -0,0 +1,382 @@
+                (1, 1),
+                activation=activation,
+            )
+            self.pre_blocks.append(c_layer)
+            tot_layers += 1
+
+        self.freq_proc_blocks = torch.nn.ModuleList([])
+        for indx in range(freq_proc_blocks):
+            c_layer = FreqWiseBlock(
+                in_channels + hid_chans * tot_layers,
+                num_freqs,
+                hid_chans,
+                activation=activation,
+            )
+            self.freq_proc_blocks.append(c_layer)
+            tot_layers += 1
+
+        self.post_blocks = torch.nn.ModuleList([])
+        for indx in range(post_blocks - 1):
+            c_layer = Conv2DActNorm(
+                in_channels + hid_chans * tot_layers,
+                hid_chans,
+                ksz,
+                (1, 1),
+                (1, 1),
+                activation=activation,
+            )
+            self.post_blocks.append(c_layer)
+            tot_layers += 1
+
+        last = Conv2DActNorm(
+            in_channels + hid_chans * tot_layers,
+            out_channels,
+            ksz,
+            (1, 1),
+            (1, 1),
+            activation=activation,
+        )
+        self.post_blocks.append(last)
+
+    def forward(self, input):
+
+        out = [input]
+        for pre_block in self.pre_blocks:
+            c_out = pre_block(torch.cat(out, 1))
+            out.append(c_out)
+
+        for freq_block in self.freq_proc_blocks:
+            c_out = freq_block(torch.cat(out, 1))
+            out.append(c_out)
+
+        for post_block in self.post_blocks:
+            c_out = post_block(torch.cat(out, 1))
+            out.append(c_out)
+
+        return c_out
+
+
+class TCNResBlock(torch.nn.Module):
+    def __init__(
+        self, in_chan, out_chan, ksz=3, stride=1, dilation=1, activation=torch.nn.ELU
+    ):
+        super(TCNResBlock, self).__init__()
+        padding = dilation
+        dconv = torch.nn.Conv1d(
+            in_chan,
+            in_chan,
+            ksz,
+            stride,
+            padding=padding,
+            dilation=dilation,
+            padding_mode="reflect",
+            groups=in_chan,
+        )
+        point_conv = torch.nn.Conv1d(in_chan, out_chan, 1)
+
+        self.layer = torch.nn.Sequential(
+            torch.nn.GroupNorm(in_chan, in_chan, eps=1e-8),
+            activations.get(activation)(),
+            dconv,
+            point_conv,
+        )
+
+    def forward(self, inp):
+        return self.layer(inp) + inp
+
+
+class TCNDenseUNet(torch.nn.Module):
+    def __init__(
+        self,
+        n_spk=1,
+        in_channels=257,
+        mic_channels=1,
+        hid_chans=32,
+        hid_chans_dense=32,
+        ksz_dense=(3, 3),
+        ksz_tcn=3,
+        tcn_repeats=4,
+        tcn_blocks=7,
+        tcn_channels=384,
+        activation=torch.nn.ELU,
+    ):
+        super(TCNDenseUNet, self).__init__()
+        self.n_spk = n_spk
+        self.in_channels = in_channels
```

Comment:
"channel" has many meanings in speech processing. We may put a comment that here channel mean num_freqs.